File: DSA/src/Questions/LinkedList/IsCycle.java
package Questions.LinkedList;

public class IsCycle {
    public class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public boolean hasCycle(ListNode head) {

        if (head == null || head.next == null) {
            return false;
        }
        ListNode slow = head;
        ListNode fast = head.next;
        while (slow.next != null && fast.next != null) {
            slow = slow.next;
            if (fast.next.next == null) {
                break;
            }
            fast = fast.next.next;
            if (slow == fast) {
                return true;
            }
        }
        return false;

    }

    public static void main(String[] args) {
        IsCycle isCycle = new IsCycle();
        ListNode head = isCycle.new ListNode(3);
        head.next = isCycle.new ListNode(2);
        head.next.next = isCycle.new ListNode(0);
        head.next.next.next = isCycle.new ListNode(-4);
        head.next.next.next.next = head.next;
        System.out.println(isCycle.hasCycle(head));
    }

}

File: DSA/Python/Questions/LinkedList/IsCycle.py
"""
# Question: Linked List Cycle
# Link: https://leetcode.com/problems/linked-list-cycle/

# Detect if a linked list has a cycle

# Time Complexity: O(n)
# Space Complexity: O(1)

# Algorithm:
# 1. Use fast and slow pointers
# 2. Fast moves two steps, slow moves one step
# 3. If pointers meet, cycle exists
# 4. If fast reaches end, no cycle

# Key Components:
# - ListNode class for linked list structure
# - has_cycle(): Main cycle detection implementation
# - Two-pointer technique (Floyd's Cycle Detection)
"""


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class IsCycle:
    def has_cycle(self, head: ListNode) -> bool:
        if not head or not head.next:
            return False

        slow = head
        fast = head.next

        while slow.next and fast.next:
            slow = slow.next
            if not fast.next.next:
                break
            fast = fast.next.next

            if slow == fast:
                return True

        return False


def main():
    solution = IsCycle()

    # Create test list with cycle: 3->2->0->-4->2
    head = ListNode(3)
    head.next = ListNode(2)
    head.next.next = ListNode(0)
    head.next.next.next = ListNode(-4)
    head.next.next.next.next = head.next  # Create cycle

    result = solution.has_cycle(head)
    print(f"Has cycle: {result}")


if __name__ == "__main__":
    main()
